[1단계 - 장바구니 미션] 니야(최수연) 미션 제출합니다.

sooyeoniya opened the pull request:
안녕하세요, 유조! 니야입니다 😄
지난 레벨 1 마지막 미션인 영화 리뷰에서 만났었는데, 이렇게 다시 뵙게 되어 반갑습니다 :)
이번에는 마지막 페어 미션인 만큼, 남은 기간 동안 잘 부탁드릴게요! 🙇🏻‍♀️

이번 미션에서는 Swagger에 명시된 실제 서버와 MSW 서버 모두 대응할 수 있도록 구성했습니다.
배포 페이지에서는 기본적으로 MSW 서버를 사용하고 있어서, http/https 네트워크 오류 문제는 발생하지 않지만,
로컬에서 직접 테스트하실 경우에는 .env 파일을 별도로 설정해주셔야 합니다!

환경 변수 구조는 `.env.example`에 예시로 넣어두었고,
로컬 테스트에 필요한 실제 .env 값은 개인 슬랙 DM으로 보내드릴게요. 확인 부탁드립니다! 🫡

## 📦 장바구니 미션

이번 미션을 통해 다음과 같은 학습 경험들을 쌓는 것을 목표로 합니다.

### 1단계

- 클라이언트 상태를 효과적으로 모델링하고 관리할 수 있다.
- Jest, React Testing Library(RTL)를 활용하여 주요 기능에 대한 테스트를 작성할 수 있다.

# 🕵️ 셀프 리뷰(Self-Review)

## 제출 전 체크 리스트

- [x] 기능 요구 사항을 모두 구현했고, 정상적으로 동작하는지 확인했나요?
- [x] RTL 테스트 케이스를 모두 작성했나요?
- [x] 배포한 데모 페이지에 정상적으로 접근할 수 있나요?

  - 배포 링크 기입: [니야의 장바구니 페이지 🛍️](https://sooyeoniya.github.io/react-shopping-cart/)

## 리뷰 요청 & 논의하고 싶은 내용

## 1) 상태 설계 의도

우선 저희는 상태 관리 과정에서 페이지 간 상태를 공유하고 Props drilling을 줄이기 위해 CartContext 라는 Context API를 만들었습니다.

먼저 상태는 크게 아래 5개로 관리하고 있습니다.

- `cartItemsData`: API에서 가져온 장바구니 상품 원본 데이터
- `cartItemsCheckData`: 각 상품의 선택 상태(체크박스) 관리
- `allChecked`: 전체 선택 버튼에 대한 상태 관리
- `isCheckDataInitialized`: 체크박스 상태가 한 번만 초기화되도록 보장하는 변수
- `errorMessage`: 에러 메시지 관리

### 1. 데이터 및 클라이언트 상태 분리

**📍 데이터 상태**

- `cartItemsData`는 SSOT 기반으로 서버와 동기화될 수 있도록 설계했습니다.

**📍 클라이언트 상태**

- `cartItemsCheckData`는 각 상품의 선택 상태(체크박스)를 관리하여, 사용자의 상호작용 결과를 표현하는 클라이언트 측 UI 상태를 관리하기 위해 추가해준 상태입니다.

### 2. 에러 처리 방식

- `errorMessage` 상태를 통해 애플리케이션 전반의 에러를 중앙 집중적으로 관리했습니다.
- 각 API 호출(fetchData, deleteItem, increaseItemQuantity 등)에서 발생한 오류를 캡처하고, 컴포넌트 계층에서는 이 `errorMessage`를 Toast로 표시하도록 하였습니다.

### 3. 체크박스 상태 초기화 로직

```tsx
useEffect(() => {
  if (cartItemsData.length > 0 && !isCheckDataInitialized.current) {
    const data = cartItemsData.map(({ id }) => ({
      id,
      checked: INITIAL_CHECKED,
    }));
    setCartItemsCheckData(data);
    isCheckDataInitialized.current = true;
  }
}, [cartItemsData]);
```

- `useRef`를 사용하여 `isCheckDataInitialized`로 체크박스 상태가 한 번만 초기화되도록 보장해주었습니다.

### 4. 파생 상태 계산

```tsx
const calculateOrderQuantity = () => { /* ... */ };
const calculateOrderPrice = () => { /* ... */ };
const calculateShippingFee = () => { /* ... */ };
```

- 원시 데이터를 기반으로 주문 수량, 가격, 배송비 등의 파생 상태를 실시간으로 계산해서 외부로 내보내주었습니다. 이들은 상태를 따로 관리하지 않고, 원시 데이터를 기반으로 계산하여, 데이터 일관성 문제가 발생하지 않도록 하였습니다.

## 2) 이번 단계에서 가장 많이 고민했던 문제와 해결 과정에서 배운 점

### ✅ 파생 상태 고민

`cartItemCheckData`는 `id`와 `checked` 속성을 가진 객체들의 배열입니다.
이 파생 상태는 **선택 여부를 독립적으로 제어하면서도 원본 데이터를 참조할 수 있도록** 설계한 구조입니다.

여기서 `checked`는 개별 상품의 선택 여부를 나타내기 위해 필수적인 속성이라고 판단했습니다.
또한 `id`는 원본 상품 데이터를 식별하기 위한 식별자로 사용되며, 해당 식별자를 기준으로 **원본 데이터에서 어떤 항목이 선택되었는지 여부(checked)를 가공하여 관리**하기 위해 포함했습니다.

`allChecked`는 전체 선택 여부를 기준으로 모든 상품의 개별 선택 상태를 일괄적으로 제어하기 위해 별개의 상태로 관리하도록 하였습니다.

## 3) 이번 리뷰를 통해 논의하고 싶은 부분

### 💬 **공통 레이아웃 분리 위치 고민**

Header와 FooterButton을 모든 페이지에서 공통으로 사용하고 있어서 최상위 컴포넌트로 분리하고 싶긴 한데, 각 도메인 영역끼리 논리적으로 묶는 것이 더 구조적으로 나아 보여서 내부에 두는 게 맞는 것 같기도 합니다. 외부로 분리하게 되면 다시 조건 분기 처리를 해줘야 해서 오히려 복잡해질 수 있다는 생각도 듭니다. 공통 컴포넌트니까 상위로 분리하는 게 좋은 건지, 아니면 각 페이지에서 직접 사용하는 방식이 더 나은지 고민입니다.

### 💬 **초기 비어있는 장바구니 테스트 필요성**

현재는 장바구니 페이지에서 시작하고 있어 장바구니에 상품이 담긴 상태를 기준으로 테스트를 진행 중입니다. 그래서 `cartItemsData`가 존재한다는 전제하에 파생 상태인 `cartItemsCheckData`도 함께 테스트하고 있습니다.

하지만 실제 사용 흐름에서는 장바구니가 비어있는 경우도 충분히 있을 수 있기 때문에, 초기 상태가 빈 장바구니일 때의 동작도 별도로 테스트하는 것이 바람직할 것 같다는 생각이 드는데, 이런 예외 상황에 대해서도 따로 처리를 해주는 것이 좋을까요?

### 💬 원본 / 파생 상태 분리 관리 필요성

현재 `cartItemsData`(원본 상태)와 `cartItemsCheckData`(파생 상태)를 하나의 Context에서 함께 관리하고 있는데요, 역할이 명확히 다른 만큼 Context를 분리하는 게 더 나을지 고민됩니다.

지금은 하나의 파일에 로직이 모두 몰려 있어 다소 복잡해 보이는데, Context를 나누면 기능별로 더 명확하게 분리되고, 유지보수도 수월해질 것 같다는 생각이 듭니다.

이런 경우, 상태를 분리해서 관리하는 방향이 더 나을까요? 아니면 연관된 데이터를 함께 관리하는 지금 방식이 더 자연스러울까요?

### 💬 페이지간 상태를 공유하는 방식

현재 `CartPage`와 `OrderPage`처럼 페이지가 나뉘어 있음에도, 상태를 모두 Context로 전역 공유하고 있습니다. 그런데 실제로는 대부분의 상태가 `CartPage`에만 필요하고, `OrderPage`에서는 특정 데이터만 넘겨받으면 되는 구조라서, **CartPage에서만 Context로 상태를 관리하고, 라우팅 시 필요한 데이터만 OrderPage에 전달하는 방식**이 더 나을까? 하는 고민이 듭니다.

전역 상태로 유지하는 편이 낫다고 볼 수도 있지만, 분리하면 페이지 간 의존도도 줄고, 상태 관리 범위도 명확해질 것 같습니다. 이런 경우, 어떤 방식이 더 적절할까요?

---

## ✅ 리뷰어 체크 포인트

<!-- 리뷰어가 이 PR을 검토할 때 중점적으로 확인할 사항입니다.
코드의 완성도뿐만 아니라, 리뷰이가 구현 과정에서 어떤 고민과 결정을 하며 학습했는지도 함께 고려해 주세요. -->

### 1. 클라이언트 상태관리

- 원본상태/파생상태를 적절히 구분하여 선언하였나요?
- React state 를 불필요하게 선언한 부분은 없나요?
- 상태 관리 로직의 책임이 적절히 응집/분리되었나요? (ex. reducer, hook, Context)

### 2. MSW/Test

- 주요 기능을 적절히 정의하였나요?
- 주요 기능/예외에 대한 테스트가 충분히 이루어졌나요?

woowa-euijinkk committed:
chore: 의존성 제거

sooyeoniya committed:
design: MobileLayout 구현

sooyeoniya committed:
chore: emotion 패키지 설치

sooyeoniya committed:
design: reset css 설정

sooyeoniya committed:
design: Header UI 구현

sooyeoniya committed:
design: FooterButton 컴포넌트 구현

sooyeoniya committed:
feat: Header의 Props 변경

Props 로 children 넘겨줘서 Header 자체를 공통 컴포넌트화

sooyeoniya committed:
feat: CartPage 구현

sooyeoniya committed:
design: Header 세로 중앙 정렬 추가

sooyeoniya committed:
design: Title 컴포넌트 구현

sooyeoniya committed:
design: Description 컴포넌트 구현

sooyeoniya committed:
design: TitleContainer 레이아웃 구현

sooyeoniya committed:
design: 전체적인 font-size 키우기

sooyeoniya committed:
design: Checkbox 구현

sooyeoniya committed:
feat: TotalSelector 컴포넌트 구현

sooyeoniya committed:
feat: QuantityCounter 컴포넌트 구현

sooyeoniya committed:
feat: CartItem 컴포넌트 구현

sooyeoniya committed:
design: Info Description 구현

sooyeoniya committed:
design: Title, Description 글씨 크기 변경

sooyeoniya committed:
feat: PriceInfo 컴포넌트 구현

sooyeoniya committed:
feat: PriceSummary 레이아웃 구성

sooyeoniya committed:
design: 내부 레이아웃 스타일 수정

sooyeoniya committed:
Merge branch 'woowacourse:main' into step1

sooyeoniya committed:
chore: npm 의존성 재설치 및 적용

sooyeoniya committed:
chore: msw 패키지 설치

sooyeoniya committed:
feat: MSW 서버 설정 및 응답 모킹

sooyeoniya committed:
chore: .gitignore에 env 추가

sooyeoniya committed:
feat: httpClient 클래스 구현

sooyeoniya committed:
feat: API 관련 상수 config 파일로 분리

sooyeoniya committed:
feat: 장바구니 GET 요청 데이터 연동

sooyeoniya committed:
feat: delete API 요청 로직 추가

sooyeoniya committed:
feat: patch API 요청 로직 추가

sooyeoniya committed:
chore: 불필요한 주석 제거

sooyeoniya committed:
feat: cartItem 삭제 기능 연동

sooyeoniya committed:
chore: Authorization에 Basic 추가

sooyeoniya committed:
feat: cartItem 수량 증감 기능 추가

sooyeoniya committed:
feat: 기본 이미지 설정

sooyeoniya committed:
design: 스크롤과 삭제 버튼 겹치지 않도록 수정

sooyeoniya committed:
chore: 페어할 때 vite dev 서버 외부 접속 가능하도록 설정

sooyeoniya committed:
design: 상품 아이템 정보들 gap 증가

sooyeoniya committed:
design: 체크 박스 active 스타일 추가

sooyeoniya committed:
feat: 현재 장바구니 상품 종류 개수 연동

sooyeoniya committed:
chore: delete와 patch 함수명 변경

복수 -> 단수

sooyeoniya committed:
feat: CartContext 구현

sooyeoniya committed:
feat: CartContext 연동

sooyeoniya committed:
chore: CartContext 내부 상태 관리 변수 및 함수명 변경

sooyeoniya committed:
feat: 전체 선택 기능 추가

sooyeoniya committed:
feat: Checkbox의 Props 타입 상속 및 onClick 추가

sooyeoniya committed:
feat: AllSelector 컴포넌트명 변경 및 전체 선택 기능 연동

sooyeoniya committed:
feat: 개별 선택 기능 추가 및 적용

sooyeoniya committed:
feat: 전체 선택 초기 상태 true로 변경

sooyeoniya committed:
chore: CartItem key 값 추가

sooyeoniya committed:
fix: 수량 증감 및 삭제 시 체크 상태 초기화 오류 수정

sooyeoniya committed:
feat: 상품 삭제 시 cartItemsCheckData 배열에 해당 상품 정보도 삭제

sooyeoniya committed:
feat: 체크 상태에 따라 주문 금액 계산 함수 구현 및 적용

sooyeoniya committed:
feat: 배송비 계산 및 적용

sooyeoniya committed:
feat: 배송비 계산 함수 외부로 분리 및 총 결제 금액 계산

- orderPrice가 0인 경우는 배송비가 0이 되도록 수정

sooyeoniya committed:
design: CartPage 레이아웃 수정

sooyeoniya committed:
feat: 장바구니에 상품이 없는 경우 Fallback UI 추가

sooyeoniya committed:
feat: 장바구니에 상품 없는 경우 주문 확인 버튼 disabled 처리

sooyeoniya committed:
feat: 장바구니 상품이 아무것도 선택되지 않은 경우 주문 확인 버튼 disabled 처리

sooyeoniya committed:
feat: 브라우저 라우팅 설정

sooyeoniya committed:
design: 주문 확인 페이지 UI 구현

sooyeoniya committed:
feat: 주문 페이지에 주문 정보 데이터 연동

sooyeoniya committed:
feat: 주문 확인 페이지에서 뒤로가기 버튼 기능 구현

sooyeoniya committed:
test: useCart 및 router 테스트 명세 작성

sooyeoniya committed:
chore: jest-fixed-jsdom 설치 및 적용

sooyeoniya committed:
test: useCart 훅 테스트 구현

sooyeoniya committed:
chore: jest-svg-transformer 설치

sooyeoniya committed:
chore: __mocks__으로 폴더명 변경

sooyeoniya committed:
chore: png -> svg 파일 확장자 변경

sooyeoniya committed:
chore: 불필요한 기존 test 파일 삭제

sooyeoniya committed:
chore: 불필요한 fileMock 파일 제거

sooyeoniya committed:
test: router 테스트 구현

sooyeoniya committed:
chore: 이름 복수형으로 변경

sooyeoniya committed:
chore: App 파일 안쓰므로 제거

sooyeoniya committed:
feat: ToastContext 기능 구현

sooyeoniya committed:
feat: Toast 기능 적용

sooyeoniya committed:
chore: tsx -> ts 파일로 확장자 변경

sooyeoniya committed:
test: ToastProvider 적용

woowa-euijinkk committed:
docs: PR template

sooyeoniya committed:
feat: Poral 추가 및 적용

sooyeoniya committed:
refactor: response 타입 수정

제네릭 제거 후 CartItems 만 정의하기

sooyeoniya committed:
chore: .env.example 추가

sooyeoniya committed:
Merge branch 'woowacourse:main' into step1

sooyeoniya committed:
feat: 상품 수량 증감 및 삭제에 대한 에러 캐치 로직 추가

sooyeoniya committed:
test: error 테스트 구현

sooyeoniya committed:
refactor: common 컴포넌트 분리

sooyeoniya committed:
chore: PropsWithChildren 임포트 방식 통일

sooyeoniya committed:
chore: 배포 스크립트 설정

sooyeoniya committed:
fix: SHOP 로고 클릭 시 루트 경로로 이동할때 경로 오류 수정

sooyeoniya committed:
chore: 불필요한 타입 제거

sooyeoniya committed:
fix: 테스트 상수 모킹 추가

sooyeoniya committed:
fix: info.svg 파일 오류 수정

yujo11 commented on src/apis/config.ts:
```diff
@@ -0,0 +1,10 @@
+    return import.meta.env.VITE_MOCK_BASE_URL;
+  }
+  return import.meta.env.VITE_API_BASE_URL;
+};
```
이런 식으로 환경 변수를 통해 관리하는 것도 방법이지만 scripts 에 분기를 넣으면 더 편하게 사용이 가능할거 같네요
ex)
- package.json
```
"scripts": {
  "dev": "vite --mode dev",
  "dev:beta": "vite --mode beta",
  "dev:mock": "vite --mode mock",
}
```

```
if (import.meta.env.MODE === 'mock') {
    return import.meta.env.VITE_MOCK_BASE_URL;
}
```

sooyeoniya replied:
유조가 추천해준 방식이, 코드 내부에서 환경변수로 분기처리하는 방식보다 각 서버 별로 테스트하기에 확실히 더 유용할 것 같다는 생각이 들었어요! 그래서 제안해주신 방향대로 한 번 수정해봤습니다.

처음에는 앞선 예시처럼 모드를 분리해서 `.env.dev`, `.env.prod`, `.env.mock` 총 3개의 env 파일을 추가로 만들었는데요.
근데 여기서 실제 서버는 하나이고 mock 서버만 다르다 보니 dev와 prod의 API URL이 동일하더라고요.

그래서 `.env` 하나에 기본 서버 API URL을 지정해두고, **mock 모드일 때만 해당 URL을 오버라이드**하도록 설정했습니다. 
덕분에 스크립트도 이전 커밋보다 더 간결해졌습니다!

- [env 및 모드 설정](https://github.com/woowacourse/react-shopping-cart/pull/337/commits/ed404370c6fb8ccc70f6782c58283d0c361c3d29)
- [스크립트 간결화](https://github.com/woowacourse/react-shopping-cart/pull/337/commits/01c13f01a0ed511368169007b58050e5cb56760e)
- [MSW 서버로 배포 가능한 스크립트 추가](https://github.com/woowacourse/react-shopping-cart/pull/337/commits/a6afacd3bbb749111e692f9bfac60a0ae9740713)

yujo11 commented on src/main.tsx:
main.tsx와 App.tsx를 분리해서 사용하지 않고 하나의 `main.tsx` 하나의 파일로 통합해서 사용하셨는데 이유가 있을까요?

sooyeoniya replied:
처음에는 `App` 컴포넌트 안에 장바구니(`Cart`) 페이지 코드만 있었는데, 이후 페이지를 분리하면서 해당 코드를 외부로 빼게 되었습니다. 그러다 보니 `App`에는 남은 코드가 없어 삭제한 채로 진행했던 것 같아요.

이후 라우팅 로직을 추가했을 때는, 프로젝트 규모가 크지 않다 보니 굳이 `main.tsx`와 `App.tsx`를 나눌 필요성을 느끼지 못했고, 전체 흐름을 한눈에 보기 쉽게 `main.tsx`에 라우팅까지 모두 작성했습니다.

하지만 다시 생각해보니, 파일이 많아질수록 **역할을 명확히 나누는 게 오히려 구조 파악에 더 도움이 되겠다**는 판단이 들었고, 최종적으로 다음과 같이 분리했습니다. https://github.com/woowacourse/react-shopping-cart/pull/337/commits/f095210c1b9489ddd31f08d113585330548a3e76

- `main.tsx`: 앱 초기화, 환경 설정(MSW 포함), 렌더링
- `App.tsx`: 라우팅 및 레이아웃 구성

분리하고 보니, 오히려 전체 흐름이 더 명확하게 보이고 관심사도 잘 나뉘어져서 구조적으로 이해하기 쉬운 코드가 된 것 같습니다!

덕분에 `main.tsx`와 `App.tsx`의 역할을 다시 한번 고민해보는 계기가 된 것 같아요! 좋은 질문 감사합니다 😄

yujo11 commented on src/main.tsx:
```diff
@@ -1,10 +1,60 @@
+  );
+};
+
+const router = createBrowserRouter(
```
`createBrowserRotuer`사용 좋네요 ㅎㅎ 👍 👍 👍

yujo11 commented on src/pages/OrderPage/OrderPage.tsx:
```diff
@@ -0,0 +1,48 @@
+
+  const navigateToBack = () => {
+    navigate(-1);
+  };
```
`/order` 경로로 유저가 바로 접근한 후에 뒤로가기를 누르면 어디로 갈까요?

sooyeoniya replied:
처음에는 배포된 페이지에서 `/order` 경로로 직접 접근했을 때 404 에러가 발생하길래, 해당 경로로는 애초에 접근이 불가능한 줄 알았습니다. 하지만 로컬 환경에서는 동일한 경로로 접근이 정상적으로 작동하길래 그 차이가 궁금해져서 원인을 찾아봤습니다!

알고 보니, 로컬에서는 Vite 개발 서버가 클라이언트 사이드 라우팅을 지원하기 때문에, 어떤 경로로 접근하더라도 자동으로 `index.html`을 반환하고 라우팅 처리를 React Router가 담당하게 됩니다.

근데, GitHub Pages는 정적 호스팅 방식이라, `/order`와 같은 경로 요청이 들어오면 해당 위치에 실제 파일이 있는지를 먼저 찾고, 없으면 404 에러를 반환하는 구조더라고요.

그래서 우선 GitHub Pages에서 SPA 라우팅을 구현하기 위해서 방법을 찾아봤는데요. ([참고1](https://github.com/orgs/community/discussions/64096), [참고2](https://www.reddit.com/r/reactjs/comments/gpj45d/how_to_handle_routes_in_github_pages/), [참고3](https://www.linkedin.com/pulse/how-make-react-routing-work-github-pages-abhishek-sharma-cjtcc/))

- HashRouter 방식
- 404 페이지 구현하여 `index.html`로 리디렉션 하는 방식

물론 Netlify나 Vercel처럼 SPA 친화적 배포 플랫폼을 사용하는 방법도 있지만, 이번 프로젝트에서는 GitHub Pages 내에서 해결해보고자 하였습니다.

두 방법 중 저는 404 페이지를 통한 리디렉션 방식을 선택했는데요. HashRouter는 경로에 `#/`이 포함되어 URL이 부자연스러워 보였고요, BrowserRouter의 일부 기능이 제한되고, SEO 측면에서도 불리하다고 하더라구요.

물론 HashRouter를 사용하면 클라이언트 사이드 라우팅이 잘 작동하고, 직접 경로 접근 시 404 에러가 발생하지 않기 때문에 이런 부분에 있어서는 좋다고 생각합니다.

하지만, 이번 미션에서는 2단계 요구사항을 봐도 애초에 사용자가 직접 URL을 입력해서 다른 경로로 이동하는 시나리오가 없습니다. (`/order` 경로는 앱 흐름 상 특정 조건을 충족했을 때만 진입해야 하는 경로) 따라서 임의 접근 시 아예 루트로 리디렉션하는 방식이 더 자연스럽고 깔끔하다고 판단했습니다!

그리고, 질문 주신 부분에 대해 고민해봤는데, 애초에 `/order` 경로로 유저가 바로 접근하는 경우에는 사용자가 방문한 이전 페이지로 돌아가기 때문에 완전 다른 웹사이트일 수도 있다고 생각했어요. 그래서 `navigate(-1)`보다는 아예 `navigate(”/”)` 로 가는 것이 더 명시적이기 때문에 훨씬 안정적이라고 판단했습니다.

- [404 페이지 추가](https://github.com/woowacourse/react-shopping-cart/pull/337/commits/240944d8710ebbade81808783dcc392c7db03524)
- [뒤로가기 버튼 경로 수정](https://github.com/woowacourse/react-shopping-cart/pull/337/commits/99b4bca6683fc1af2f454484d80c75169a9cdd4a)

yujo11 replied:
오 호스팅 방식까지 의도하고 질문 드린건 아니었는데 그 부분까지 찾아보셨군요 멋집니다 ㅎㅎ

이 문제는 다른 플랫폼을 이용해서 배포하는 경우(ex - AWS S3 + CloudFront)에도 발생할 수 있는 문제로 보이는데요. 플랫폼의 특성을 넘어 SPA가 어떻게 파일들을 불러오고, 렌더링 되어 유저에게 보여지는지 그 과정에 대해서도 한번 생각해보시면 좋을거 같네요

sooyeoniya replied:
오 좋은 정보 감사합니다! 꼭 찾아보면서 공부하겠습니다!! 🙇🏻‍♀️

yujo11 commented on src/pages/CartPage/CartPage.tsx:
```diff
@@ -0,0 +1,92 @@
+  const navigate = useNavigate();
+
+  const navigateToOrderPage = () => {
+    navigate("/order");
```
이런 path들은 상수화를 해서 사용해주셔도 좋을거 같아요

sooyeoniya replied:
라우팅 경로들 상수화 하는 것에 동의합니다! 훨씬 유지보수 관점에서 관리하기 수월할 것 같아요. 반영했습니다😃 https://github.com/woowacourse/react-shopping-cart/pull/337/commits/7bee46a5710d1d21795901f175247248df12a6a9

yujo11 commented on src/contexts/ToastContext.tsx:
```diff
@@ -0,0 +1,71 @@
+  useEffect(() => {
+    return () => {
+      if (toastTimer.current) {
+        clearTimeout(toastTimer.current);
```
타이머 클리어를 꼼꼼하게 잘 해주셨네요 ㅎㅎ 👍 👍 👍

yujo11 commented on src/contexts/ToastContext.tsx:
```diff
@@ -0,0 +1,71 @@
+
+export const ToastProvider = ({ children }: PropsWithChildren) => {
+  const [message, setMessage] = useState("");
+  const [type, setType] = useState<ToastType>(TOAST_TYPES.SUCCESS);
+  const toastTimer = useRef<ReturnType<typeof setTimeout> | null>(null);
```
`useState`로 선언한 변수와 `useRef`로 선언한 변수는 어떤 차이가 있나요?

sooyeoniya replied:
`useState`와 `useRef`의 가장 큰 차이는 리렌더링 트리거 여부랑 제어와 비제어 관점에서의 차이라고 생각합니다!

`useState`는 React가 직접 관리하는 제어 상태이기 때문에, 상태가 변경되면 해당 state를 참조하는 컴포넌트가 자동으로 리렌더링 됩니다. 예를 들어, 제 Toast 코드에 있는 message나 type과 같이, 값이 바뀌었을 때 UI에 직접 영향을 주어야 하는 값들은 useState로 관리해야합니다.

`useRef`의 경우에는 비제어 상태로, ref 객체의 current 프로퍼티로 값을 관리하게 됩니다. 이때 current를 직접 수정해도 객체 참조값이 바뀌지 않기 때문에 리렌더링이 발생하지 않습니다. 그래서 지금 코드에서도 보시다시피, setTimeout의 반환값인 timeoutID를 useRef로 관리한 이유는 이 값 자체가 컴포넌트를 렌더링할 필요가 없고, 내부적으로만 관리하면 되는 값이기 때문입니다. 만약 이 값을 useState로 관리했다면, 값이 바뀔 때마다 불필요한 렌더링이 발생합니다!

yujo11 commented on src/pages/CartPage/CartPage.tsx:
```diff
@@ -0,0 +1,92 @@
+              <AllSelector />
+              <S.CartItemsContainer>
+                {cartItemsData.map((cartItem) => (
+                  <CartItem key={cartItem.id} cartItem={cartItem} />
```
React에서 key는 어떤 역할을 하나요?

sooyeoniya replied:
React에서 key는 리스트 같은 반복적인 요소를 렌더링 할 때 각 요소를 고유하게 식별하기 위한 값입니다. 예를 들어 map으로 배열을 렌더링할 때, React는 어떤 항목이 변경, 추가, 삭제됐는지를 key를 기준으로 판단하게 됩니다.

여기서 단순히 index를 key로 사용하면 문제가 생길 수 있는데요, index는 요소의 위치에 따라서 결정되기 때문에, 중간에 항목이 삽입되거나 삭제될 경우, 이후 항목들의 key 값이 하나씩 밀리거나 당겨져서 모두 바뀌게 되고, React는 이들을 모두 새로운 요소로 인식해 불필요한 리렌더링이 발생하게 됩니다. 따라서, 지금 코드를 보면 cartItemsData는 정적 데이터가 아니라 동적으로 변동 가능한 데이터이기 때문에, 이런 경우에는 index를 key로 사용하기에는 적합하지 않다고 판단했습니다.

그래서 cartItem.id(서버에서 받아온 primary key)로 key 값을 설정해주었는데요, cartItem.id처럼 각 항목을 고유하게 식별할 수 있는 값을 key로 사용하면, 중간에 변경이 생겨도 해당 항목만 리렌더링되고 나머지는 그대로 유지됩니다. 이렇게 하면 React 렌더 단계의 재조정 과정에서도 훨씬 성능도 향상되고, 예상치 못한 UI 버그도 막을 수 있습니다!

마침 이번 주에 저희 크루 `아더`가 이 주제로 테코톡 발표를 해서 더 기억에 남았던 것 같아요. 덕분에 복습도 되고 좋았습니다! 감사합니다 :)

yujo11 commented on src/pages/CartPage/CartPage.tsx:
```diff
@@ -0,0 +1,92 @@
+              </S.CartItemsContainer>
+            </S.CartContainer>
+            <S.InfoContainer>
+              <img src={InfoIcon} alt="info" />
```
오 alt 태그도 잘 챙겨주셨군요 ㅎㅎ 👍 👍 👍

yujo11 commented on src/contexts/ToastContext.tsx:
```diff
@@ -0,0 +1,71 @@
+    <ToastContext.Provider value={{ showToast }}>
+      {children}
+      {!!message && (
+        <Portal>
```
Portal을 사용하신 이유가 있나요?

sooyeoniya replied:
Portal을 사용한 이유는 쌓임 맥락을 직접 제어하기 위해서였습니다.
아래 이미지를 보면, Portal 사용 전과 후의 차이를 쉽게 이해할 수 있습니다!

<img width="1800" alt="before" src="https://github.com/user-attachments/assets/25955d33-02e4-428a-9cf0-73574cc50326" />

![after](https://github.com/user-attachments/assets/5b80ef80-cc03-41dd-bf01-3294f299a6c7)

 처음에는 각 페이지의 Main 레이아웃에 id를 추가해 Portal의 마운트 위치를 지정하려고 했습니다. 근데 실제 렌더링 흐름에서는 ToastProvider 내부에서 Portal이 먼저 렌더링되기 때문에, 원하는 시점에 custom-root라는 id가 생성되지 않아 결국 document.body에 마운트되는 문제가 발생했습니다.

그래서 해결하기 위해 MobileLayout에서 custom-root id를 추가해주었습니다.

그리고 Toast가 해당 영역 내에서 제대로 위치하도록 하기 위해, custom-root를 감싸고 있는 Content 컴포넌트에 `position: relative` 속성을 부여해 포지셔닝 기준을 명확히 설정해주었습니다. 이로 인해 Toast가 예상한 위치에 정상적으로 표시되도록 할 수 있었습니다!

yujo11 commented on src/contexts/CartContext.tsx:
```diff
@@ -0,0 +1,228 @@
+  >([]);
+  const [allChecked, setAllChecked] = useState(INITIAL_CHECKED);
+  const isCheckDataInitialized = useRef(false);
+  const [errorMessage, setErrorMessage] = useState("");
```
단순히 `errorMessage`만 외부로 노출해주고 있고, 이렇게 노출된 errorMessage는 showToast를 통해 메세지를 노출해주는 용도로만 처리 되고 있는데요. 그렇다면 showToast까지 이 hook 안에서 처리하는 방법도 있을거 같아요. 이렇게 구현하신 배경이 궁금합니다!

sooyeoniya replied:
처음에는 데이터 처리 로직은 CartContext에서 담당하고, UI 표시 방식은 UI 레이어에서 결정하도록 설계해 관심사를 분리하고자 했습니다. 또한, Context가 특정 UI 요소(Toast 컴포넌트)에 직접적으로 의존하지 않도록 함으로써, 의존성 방향이 한쪽으로만 흐르도록 유지하려는 목적도 있었습니다!

그래서 errorMessage 상태는 여러 컴포넌트에서 공통으로 사용할 수 있도록 훅 내부에서 관리하고, showToast() 호출은 이 상태를 소비하는 컴포넌트 쪽의 책임으로 넘겼던 것 같아요.

하지만 다시 생각해보니, 에러 메시지와 토스트 호출이 분리되어 있다 보니 응집도가 떨어지고 사용성도 불편한 구조라는 판단이 들었습니다. 게다가 실제로 토스트는 ToastProvider 내에서 독립적으로 렌더링되기 때문에, 굳이 CartPage 같은 특정 컴포넌트에 의존할 필요도 없다는 점을 깨달았어요. 그래서 에러 메시지 관리와 토스트 호출을 한 곳에서 함께 처리하는 방식으로 코드 수정했습니다! https://github.com/woowacourse/react-shopping-cart/pull/337/commits/987145fc7530d695c8e4f8b91a2e1b4ca50f6051

다만 이 과정에서 또 다른 고민이 생겼는데요, 에러 상태 관리와 에러 토스트 처리를 완전히 분리하기 위해 ErrorContext를 별도로 만들까도 고민했습니다. 이렇게 하면 에러 상태만 따로 관리할 수 있어서 오히려 관심사 분리가 더 명확해질 수 있다고 생각했어요.

하지만 ErrorContext 내부에서 useToast()를 사용해야 하다 보니 결국 ToastContext에 의존하게 되고, 또 에러를 발생시키는 로직(CartContext)과 이 에러 상태를 표시하는 로직(ErrorContext)이 따로 분리해 존재하는 구조 자체도 오히려 응집도를 해칠 수 있다는 고민이 들었습니다.

결국, 현재는 에러 발생과 처리(토스트 호출)를 같은 위치에서 함께 관리하는 방향으로 정리했습니다. 
다만 CartContext에 너무 많은 책임이 몰려 있는 구조여서, 역할 분리에 대해 고민이 좀 더 필요할 것 같습니다!

yujo11 commented on src/__tests__/useCart.test.tsx:
```diff
@@ -0,0 +1,364 @@
+    (getCartItems as jest.Mock).mockResolvedValue(mockCartItems);
+  });
+
+  describe("초기 렌더링했을 때", () => {
```
context를 잘 나눠주셨네요 👍 👍 👍

yujo11 commented on src/__tests__/useCart.test.tsx:
```diff
@@ -0,0 +1,364 @@
+  });
+
+  describe("초기 렌더링했을 때", () => {
+    it("전체 장바구니 상품 데이터가 불러와지고, 원본 상태와 파생 상태가 초기화된다.", async () => {
```
`원본 상태`, `파생 상태` 가 각각 무엇인지 알 수 있게 명시해주면 좋을거 같아요

sooyeoniya replied:
원래는 `cartItemsData`가 원본 상태이고, `cartItemsCheckData`는 거기서 파생된 상태라고 생각해서 테스트 명세를 작성했었는데요. 다시 생각해보니까, 파생 상태라는 건 원본 데이터를 가공해서 나온 상태를 의미하잖아요? 그런 기준에서 보면 cartItemsCheckData는 사실 파생이라기보단 원본에 가까운 것 같더라고요. 왜냐하면 이 데이터는 `{ id, checked }` 형태로 되어 있는데, id는 원본 데이터를 그대로 가져온 거고, checked는 클라이언트에서 별도로 만들어낸 상태이기 때문입니다.

```tsx
interface CartItemCheckType {
  id: number;
  checked: boolean;
}
```

그래서 이 기준에 맞춰 다시 상태를 분류해봤습니다.

**원본 상태**: `cartItemsData`(서버에서 받은 장바구니 데이터), `cartItemsCheckData`(원본의 id를 그대로 사용하고, checked만 추가), allChecked(전체 선택 여부)

**파생 상태**: `cartItemCount`(장바구니 상품 개수), `orderItemCount`(주문할 상품 종류 개수), `orderQuantity`(주문할 상품 수량), `orderPrice`(주문할 상품 가격), `shippingFee`(배송비), `totalPrice`(총 결제 금액: orderPrice + shippingFee)

그리고 이런 상태들이 잘 초기화되는지를 확인하기 위해 테스트 코드를 수정했습니다. 모든 파생 상태를 다 테스트하는 것이 아니라, 실제 요구사항에서 중요한 값들 위주로 검증했습니다. 예를 들어 totalPrice나, 무료 배송 적용 기준 같은 로직은 중요한 경계값이 있는 부분이라 테스트에 포함했습니다.

또 파생 상태가 많다보니, 아예 원본 상태와 파생 상태를 나누어서 테스트하도록 수정했습니다! https://github.com/woowacourse/react-shopping-cart/pull/337/commits/1d90d3f9c429e5ab8350f8950d39e3254b4fa9ba

yujo11 commented on src/contexts/CartContext.tsx:
여기에서 데이터 관리, API 통신 등 굉장히 많은 일들을 한번에 수행하고 있는데요. 너무 많은 책임과 역할을 이 Context가 가져가는거 같아요. 로직을 책임 단위로 나눠 볼 수 있을까요?

sooyeoniya replied:
말씀해주신 것처럼 `CartContext`가 많은 일을 하고 있기 때문에, 내부 로직에서 커스텀 훅으로 나누어서 관리해보도록 했습니다! https://github.com/woowacourse/react-shopping-cart/pull/337/commits/d5c84e297b29bc61c94c0158d70d161cddb49582

전반적으로 컴포지션 패턴을 좀 활용해서 각 훅이 단일 책임에 가깝게 구현되어있으면서 서로 조합해서 기능을 구현할 수 있도록 했습니다. 덕분이 관심사 분리도 잘 되어 유지보수하기 편리해진 것 같습니다! 다만, 한 가지 고민인 점은 컴포지션 패턴이다보니 서로 의존을 많이 하고 있어서 결합도가 좀 높다는 문제가 있긴 하네요. 🥲

`CartProvider`에서 사용하는 커스텀 훅을 한 번 요약해봤습니다!

- `useCartData`: 장바구니 아이템 관련 API 호출 및 데이터 상태 관리
- `useCartCheck`: 장바구니 아이템 선택 상태 관리 훅(전체 선택 및 해제, 개별 아이템 선택 기능 등)
- `useCartCalculation`: 선택된 장바구니 아이템에 대한 수량, 가격, 배송비, 총액 등을 계산하는 훅

앞서 [errorMessage와 관련된 피드백](https://github.com/woowacourse/react-shopping-cart/pull/337#discussion_r2113843159)에서도 말씀드렸듯이, 에러 처리 관련 관심사 분리에 대한 고민을 좀 더 해봤는데요. `useErrorHandler`라는 훅을 새로 만들어서 에러 처리 로직을 집중적으로 담당해주는 훅을 만들었습니다!

에러 처리 관련해서 훅 관계도도 한눈에 보기 쉽게 정리해봤습니다.

`ToastContext → useToast → useErrorHandler → useCartData → CartContext`

- `ToastContext`: 토스트 메시지 상태 관리 및 표시를 위한 컨텍스트
- `useToast`: ToastContext를 사용해 토스트 메시지 표시 기능을 해주는 훅
- `useErrorHandler`: useToast를 사용해 에러 메시지를 토스트로 표시하는 기능을 캡슐화
- `useCartData`: useErrorHandler를 사용해 API 호출 중 발생하는 에러 처리
- `CartContext`: 장바구니 관련 로직 제공 컨텍스트

yujo11 replied:
hook을 잘 나눠주신거 같네요 멋집니다 니야 ㅎㅎ 👍 👍 👍

yujo11 commented on src/main.tsx:
<img width="251" alt="image" src="https://github.com/user-attachments/assets/73b249ad-9aa6-4174-83f9-b369c155f18c" />

상품이 일정 영역에서만 렌더링 되고 잘리는데 이 부분은 의도된 동작일까요?

sooyeoniya replied:
의도한 부분은 아니었고, 화면을 축소해서 테스트해보는 걸 미처 놓쳤네요. 알려주신 덕분에 내부 목록들이 화면 비율에 맞게 꽉 차도록 수정했습니다! 감사합니다 😄  https://github.com/woowacourse/react-shopping-cart/pull/337/commits/0cec00a3d93696615d985bbd6223e19ea71b91ae

<img width="1800" alt="image" src="https://github.com/user-attachments/assets/4751fe64-ce80-4485-ab58-824318b89456" />

yujo11 reviewed (changes requested):
안녕하세요 니야 ~ 다시 만나서 반갑습니다 ㅎㅎ 벌써 마지막 미션이네요 이번 미션도 잘 부탁드려요 !

전체적으로 일관성 있는 컨벤션을 바탕으로 코드를 잘 작성해주셔서 남길 코멘트가 많지 않았는데요. 코드를 보면서 궁금했던 부분들 위주로 코멘트를 남겼으니 확인해주시면 좋을거 같네요!

>💬 공통 레이아웃 분리 위치 고민

최상위 컴포넌트가 `Layout`을 말씀하시는거라면 지금과 같은 구조가 더 자연스러울거 같은데요. 현재 2개의 페이지만 존재하는 상황에서도 Header와 Footer의 동작이 각각 다르기 때문에 상위 컴포넌트에서 이에 대한 조건을 분기 처리하는 것보다 상황에 맞게 각 페이지가 책임을 가져가는 것이 좋을거 같아요

>💬 초기 비어있는 장바구니 테스트 필요성

"장바구니가 비어 있는 상태"는 예외 상황이 아니라 일반적인 시나리오로 보이는데요.이런 일반적인 시나리오에 대해서는 테스트를 진행해주는 것이 좋다고 생각합니다

>💬 원본 / 파생 상태 분리 관리 필요성

파생 상태가 원본 상태에 완전히 의존하고 있기 때문에 현재처럼 응집도 있게 두는 방향도 나쁘지 않게 느껴져요. 다만 코멘트로도 남겼듯이 현재 CartContext에서 너무 많은 일들을 하고 있는데요. 내부 로직을 적절히 커스텀 훅으로 나눠보는 방법도 생각해 볼 수 있을거 같아요

>💬 페이지간 상태를 공유하는 방식

상황마다 다를거 같은데요. 현재처럼 전역상태를 유지하는 방법에도 장단점이 존재하고 페이지 라우팅 시 상태를 공유하는 방법에도 장단점이 존재하기 때문에 이를 먼저 고려해보면 좋을거 같아요.

현재는 장점에 대한 부분만 서술해주셨는데 각 방법의 장단점은 어떤게 있을까요? 그리고 현재 구조에서 어떤 방법을 사용하는게 더 적절할까요? 정답은 없는 부분이지만 이 부분은 니야가 조금 더 고민을 해봐도 좋을 부분이라고 생각합니다 ㅎㅎ

---


미션 진행하느라 고생 많으셨습니다 니야 ~~ 👍 👍 👍

sooyeoniya committed:
chore: api 로직 컨벤션 및 변수명 수정

sooyeoniya committed:
fix: API 환경 변수 설정 개선 및 중복 제거로 요청 URL 문제 해결

환경변수로 소스코드에서 getBaseUrl 함수를 통해 분기처리하던 로직을 아예 환경변수 파일을 분리하여 스크립트로 나눠서 적용
MODE가 mock일 때만 MSW 서버 실행하도록 함

sooyeoniya committed:
refactor: dev, prod 모드 환경변수가 동일하므로, 기본 env에 정의하고 mock 모드일 때만 오버라이드되도록 수정

sooyeoniya committed:
refactor: App.tsx, main.tsx 파일 나누어 관심사 분리

sooyeoniya committed:
chore: 배포 시 MSW 모킹 서버로 배포 가능한 스크립트 추가

sooyeoniya committed:
fix: 화면 높이에 맞게 장바구니 목록을 동적으로 모두 채워지도록 변경

- Main의 height 지정 -> header와 footerbutton 의 높이 뺀 값
- ContentContainer 는 스크롤 되지 않도록 overflow: hidden 적용
- CartContainer에 min-height: 0 적용 -> 이를 적용하지 않으면 flex 내부 항목들이 콘텐츠에 따라서 계속 확장됨

sooyeoniya committed:
refactor: Route Path 상수화

sooyeoniya committed:
test: 파생 상태에 대한 테스트 추가 구현

sooyeoniya committed:
refactor: 에러 표시 로직 호출 위치 수정

- CartPage -> CartContext

sooyeoniya committed:
chore: Order 페이지에서 뒤로가기 버튼 클릭 시 '/' 경로로 명시적으로 이동하도록 수정

sooyeoniya committed:
feat: 404 페이지 구현하여 루트로 리디렉션하기

sooyeoniya committed:
test: 장바구니가 빈 경우 테스트 구현

sooyeoniya committed:
chore: errorMessage Context value에서 제거

sooyeoniya commented:
> "장바구니가 비어 있는 상태"는 예외 상황이 아니라 일반적인 시나리오로 보이는데요.이런 일반적인 시나리오에 대해서는 테스트를 진행해주는 것이 좋다고 생각합니다
> 

생각해보니까 당연히 장바구니 빈 경우도 일반 시나리오네요! 장바구니 빈 경우에 대해서도 테스트 추가했습니다 😃 https://github.com/woowacourse/react-shopping-cart/pull/337/commits/ecb262ab4018dba385b2503ecee9c175ff3ba882

sooyeoniya commented:
> 상황마다 다를거 같은데요. 현재처럼 전역상태를 유지하는 방법에도 장단점이 존재하고 페이지 라우팅 시 상태를 공유하는 방법에도 장단점이 존재하기 때문에 이를 먼저 고려해보면 좋을거 같아요.
> 

상황에 따라 달라질 수 있다는 점에 공감합니다!

현재처럼 전역 상태로 데이터를 관리하는 방식은 페이지 간의 의존도를 줄이고, 상태 관리 측면에서도 훨씬 수월하다는 장점이 있는 것 같습니다. 다만, `OrderPage`에서는 필요한 데이터만 넘겨받으면 되는 구조임에도 불구하고, 전역 상태에 의존하다 보니 불필요한 데이터까지 함께 관리하게 되는 문제가 있습니다.

물론, 라우팅 시마다 필요한 데이터를 페이지 단위로 넘기는 방식도 고려해봤지만, 
2단계 미션을 보면 지금보다 훨씬 다루어야 할 데이터가 많아지기 때문에, 오히려 관리가 더 복잡해질 수 있다고 판단했습니다.

따라서 우선은 지금처럼 전역 상태로 관리하되, 전역 상태 내부를 도메인별로 나눠 좀 더 구조적으로 관리하는 방향으로 진행해보려고 합니다!

sooyeoniya committed:
refactor: useErrorHandler 커스텀훅 구현 및 적용

sooyeoniya committed:
refactor: CartContext 내부 로직 관심사별 커스텀훅으로 분리

sooyeoniya commented:
안녕하세요, 유조! 니야입니다.
장바구니 1단계 미션 반영했습니다 :)
덕분에 고민할 거리도 많이 생겨서, 질문 하나하나 곱씹으면서 여러 방향으로 생각해볼 수 있었던 것 같아요.
질문 주신 부분에 대해서는 각각 답변을 달아두었으니, 확인 부탁드립니다! 🙇🏻‍♀️

yujo11 commented on package.json:
```diff
@@ -4,17 +4,24 @@
   "scripts": {
-    "dev": "vite",
+    "dev": "vite dev --host 0.0.0.0",
+    "dev:mock": "vite dev --host 0.0.0.0 --mode mock",
```
👍 👍 👍

yujo11 reviewed (approved):
안녕하세요 니야

코멘트들 꼼꼼하게 확인하고 더 나아가 많은 고민을 하시면서 코드를 수정해주신 부분들이 보여 인상 깊었네요 ㅎㅎ 👍 👍 👍 💯 💯 

소소한 코멘트 추가로 남긴 부분이 있는데 확인해보시면 좋을거 같습니다! 고생 많으셨습니다 니야 ~~ 👍 👍 👍
